### PR TITLE
Add a Django example and setup.py

### PR DESCRIPTION
**Pull request by LiuG-lynx**
I've added examples of the Django framework to this repository, check out the README to learn more.There is currently no packaged version.It can only be used as a local py file @gerbyzation

I am not sure how to package this Repository currently. I have added setup.py but it has not been tested. Where can I get help @hsluoyz

### CODE REVIEW

**Comment by LiuG-lynx** [comment]
It can be viewed in this repository before  merging @gerbyzation
[https://github.com/LiuG-lynx/postgresql-watcher](url)

**Comment by hsluoyz** [comment]
@LiuG-lynx we can review your code here in this PR. This is why we have Pull Request :) I just want to make sure the time that you make it ready for review.

**Comment by LiuG-lynx** [comment]
@hsluoyz Please refer to the link text to set the username and password, and trigger the github action when a new release is built. Since I am not sure how to create a fake postgresql connection, I did not verify the unit test on the github action, only verified it locally.
https://docs.github.com/en/actions/reference/encrypted-secrets

**Comment by hsluoyz** [comment]
@gerbyzation @techoner plz review.

**Comment by leeqvip** [comment]
@LiuG-lynx Can you squash your commits into one?

File: postgresql_watcher/watcher.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+            print("Casbin policy update identified..")
+            conn.poll()
+            while conn.notifies:
+                notify = conn.notifies.pop(0)
+                print(f"Notify: {notify.payload}")
+
+
+
+
+class PostgresqlWatcher(object):
```
Comment (gerbyzation) [inline postgresql_watcher/watcher.py:73]: I think this is supposed to extend the ABC defined above?

File: examples/django_examples/README.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    'casbin_middleware.middleware.CasbinMiddleware', # Add this line, must after AuthenticationMiddleware.
+]
+```
+
+- Copy ``casbin_middleware`` folder to your Django's top folder, modify ``casbin_middleware/middleware.py`` if you need:
+
+```python
+import casbin
+
+    def __init__(self, get_response):
```
Comment (gerbyzation) [inline examples/django_examples/README.md:37]: Looks like this is missing the `class` definition

File: postgresql_watcher/watcher.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        Method which checks if there is an update necessary for the casbin
+        roles. This is called with each flask request.
+        Returns:
+            Bool
+        """
+        pass
+
+
+def casbin_subscription(
+    process_conn:Any,
```
Comment (gerbyzation) [inline postgresql_watcher/watcher.py:46]: would be nice to have a type here.

File: postgresql_watcher/watcher.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
+from typing import Optional, Any
+from psycopg2 import connect, extensions
+from multiprocessing import Process, Pipe
+import time
+from select import select
```
Comment (gerbyzation) [inline postgresql_watcher/watcher.py:5]: Where does this come from, I don't see it listed in `requirements`?

File: postgresql_watcher/watcher.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    curs.execute(f"LISTEN {channel_name};")
+    print("Waiting for casbin policy update")
+    while True and not curs.closed:
+        if not select([conn], [], [], 5) == ([], [], []):
+            print("Casbin policy update identified..")
+            conn.poll()
+            while conn.notifies:
+                notify = conn.notifies.pop(0)
+                print(f"Notify: {notify.payload}")
+
```
Comment (gerbyzation) [inline postgresql_watcher/watcher.py:69]: This doesn't seem to call anything when it's notified of a change?

File: postgresql_watcher/watcher.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            if self.parent_conn.poll():
+                message = self.parent_conn.recv()
+                print(f"message:{message}")
+                return True
+        except EOFError:
+            print(
+                "Child casbin-watcher subscribe process has stopped, "
+                "attempting to recreate the process in 10 seconds..."
+            )
+            self.subscribed_process, self.parent_conn = self.create_subscriber_process(
```
Comment (gerbyzation) [inline postgresql_watcher/watcher.py:152]: this is creating the subscriber process twice (first time on line 88)?

File: tests/test_postgresql_watcher.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    pg_watcher = PostgresqlWatcher(host=HOST, port=PORT, user=USER, password=PASSWORD)
+    return pg_watcher
+
+
+pg_watcher = get_watcher()
+
+
+class TestConfig(unittest.TestCase):
+    def test_pg_watcher_init(self):
+        assert isinstance(pg_watcher.parent_conn, connection.PipeConnection)
```
Comment (gerbyzation) [inline tests/test_postgresql_watcher.py:22]: Based on the unittest output I think this should be `connection.Connection`
```
======================================================================
ERROR: test_pg_watcher_init (__main__.TestConfig)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_postgresql_watcher.py", line 26, in test_pg_watcher_init
    assert isinstance(pg_watcher.parent_conn, connection.PipeConnection)
AttributeError: module 'multiprocessing.connection' has no attribute 'PipeConnection'
```

**Review by gerbyzation — CHANGES_REQUESTED** [review]
@LiuG-lynx thanks for the effort you're putting in. Unfortunately the tests fail for me when running locally (see my other comment for the error). I've had a look through the code and left some first reactions. 

I think something that would be worth doing is improving the testing a bit further, would be great to have more of it's functionality tested and will help in getting this working. I added this to test if two enforcers would sync (needs proper teardown of connections/processes):
```
import time
import casbin
import casbin_sqlalchemy_adapter

class TestFunctionality(unittest.TestCase):
    def setUp(self):
        adapter = casbin_sqlalchemy_adapter.Adapter(
            "postgres://postgres:postgres@localhost:5432/postgres"
        )
        enforcer = casbin.Enforcer("examples/rbac_model.conf", adapter)
        enforcer.remove_policy("group1", "target3", "read")

    def test_enforcers_sync(self):
        adapter = casbin_sqlalchemy_adapter.Adapter(
            "postgres://postgres:postgres@localhost:5432/postgres"
        )
        enforcer_1 = casbin.Enforcer("examples/rbac_model.conf", adapter)
        watcher_1 = get_watcher()
        watcher_1.set_update_callback(enforcer_1.load_policy)
        enforcer_1.set_watcher(watcher_1)

        adapter_2 = casbin_sqlalchemy_adapter.Adapter(
            "postgres://postgres:postgres@localhost:5432/postgres"
        )
        enforcer_2 = casbin.Enforcer("examples/rbac_model.conf", adapter_2)
        watcher_2 = get_watcher()
        watcher_2.set_update_callback(enforcer_2.load_policy)
        enforcer_2.set_watcher(watcher_2)

        enforcer_1.add_policy("group1", "target3", "read")
        assert enforcer_1.enforce("group1", "target3", "read") is True
        # give it a sec to reconcile
        time.sleep(1)
        assert enforcer_2.enforce("group1", "target3", "read") is True
```
This test fails for me at the moment, it's bedtime here now so I can look into why it doesn't sync yet another time if needed.

I think we can make the examples clearer. At the moment it explains how to integrate with flask and django which I think add noise as integration (generally) happends with the enforcer and not with the web server. The django example uses CSV as model storage, sqlite for the Django ORM and this package and postgres to sync - not easy to follow what is going on. I'd suggest using a simple setup for the primary examples (`casbin_sqlalchemy_adapter` for example) and if needed provide documentation for integration with web framework/casbin integration packages in a less prominent place.

For spinning up a postgres database locally you can use docker:
```
docker run --name casbin-watcher-postgres -p 5432:5432 -e POSTGRES_PASSWORD=postgres -d postgres
```
I can help with setting up postgres with docker in Github Actions.

**Comment by hsluoyz** [comment]
@LiuG-lynx plz resolve the conflicts and respond to reviews.

image

**Comment by Zxilly** [comment]
@LiuG-lynx any progress on this?

**Comment by hsluoyz** [comment]
@Zxilly he already quited. So no reply would be expected. Can you borrow the useful code in this PR and make a new PR instead?

**Pull request closed without merging**